File: android/app/src/main/java/com/stepserve/app/data/api/ApiService.kt
package com.stepserve.app.data.api

import okhttp3.MultipartBody
import retrofit2.Response
import retrofit2.http.*

interface ApiService {

    // ── Auth ─────────────────────────────────────────────────────────────────

    @POST("auth/register")
    suspend fun register(@Body body: RegisterRequest): Response<TokenResponse>

    @POST("auth/login")
    suspend fun login(@Body body: LoginRequest): Response<TokenResponse>

    @GET("auth/me")
    suspend fun me(): Response<UserMe>

    @POST("auth/forgot-password")
    suspend fun forgotPassword(@Body body: ForgotPasswordRequest): Response<MessageResponse>

    @POST("auth/reset-password")
    suspend fun resetPassword(@Body body: ResetPasswordRequest): Response<MessageResponse>

    // ── Public marketplace ───────────────────────────────────────────────────

    @GET("stepserve/home")
    suspend fun home(): Response<HomeData>

    @GET("categories")
    suspend fun categories(): Response<List<Category>>

    @GET("search/services")
    suspend fun searchServices(
        @Query("query") query: String? = null,
        @Query("category_id") categoryId: Int? = null,
        @Query("location") location: String? = null,
        @Query("min_price") minPrice: Double? = null,
        @Query("max_price") maxPrice: Double? = null,
        @Query("limit") limit: Int = 50,
        @Query("offset") offset: Int = 0,
    ): Response<List<Service>>

    @GET("services")
    suspend fun listServices(): Response<List<Service>>

    // ── Provider profile ─────────────────────────────────────────────────────

    @POST("providers/profile")
    suspend fun upsertProfile(@Body body: ProviderProfileUpsert): Response<ProviderProfile>

    // ── Provider uploads ──────────────────────────────────────────────────────

    @Multipart
    @POST("providers/uploads")
    suspend fun uploadFile(@Part file: MultipartBody.Part): Response<ProviderUpload>

    @GET("providers/uploads")
    suspend fun listUploads(): Response<List<ProviderUpload>>

    // ── Services (provider) ──────────────────────────────────────────────────

    @POST("services")
    suspend fun createService(@Body body: ServiceCreate): Response<Service>

    @PATCH("services/{service_id}")
    suspend fun updateService(
        @Path("service_id") serviceId: Int,
        @Body body: ServiceUpdate,
    ): Response<Service>

    @DELETE("services/{service_id}")
    suspend fun deactivateService(@Path("service_id") serviceId: Int): Response<Map<String, Any>>

    @GET("provider/listings")
    suspend fun providerListings(): Response<List<Service>>

    @GET("provider/dashboard")
    suspend fun providerDashboard(): Response<ProviderDashboard>

    // ── Listing payment ───────────────────────────────────────────────────────

    @POST("listings/{service_id}/pay")
    suspend fun payListing(@Path("service_id") serviceId: Int): Response<ListingPaymentResponse>

    // ── Bookings ─────────────────────────────────────────────────────────────

    @POST("bookings")
    suspend fun createBooking(@Body body: BookingCreate): Response<BookingResponse>

    @GET("customer/bookings")
    suspend fun customerBookings(): Response<List<CustomerBooking>>

    // ── Payments ─────────────────────────────────────────────────────────────

    @POST("payments")
    suspend fun createPayment(@Body body: PaymentCreate): Response<PaymentResponse>

    // ── Reviews ──────────────────────────────────────────────────────────────

    @POST("reviews")
    suspend fun createReview(@Body body: ReviewCreate): Response<Map<String, Any>>

    // ── Admin ─────────────────────────────────────────────────────────────────

    @GET("admin/overview")
    suspend fun adminOverview(): Response<AdminOverview>

    @GET("admin/users")
    suspend fun adminUsers(): Response<List<AdminUser>>

    @PATCH("admin/users/{user_id}/status")
    suspend fun adminUpdateUserStatus(
        @Path("user_id") userId: Int,
        @Query("active") active: Boolean,
    ): Response<Map<String, Any>>

    @GET("admin/bookings")
    suspend fun adminBookings(): Response<List<AdminBooking>>

    @GET("admin/services")
    suspend fun adminServices(): Response<List<Service>>

    @POST("categories")
    suspend fun createCategory(@Body body: CategoryCreate): Response<Category>
}

File: ios/StepServe/Data/APIService.swift
import Foundation

// All API calls as static async methods, using NetworkClient.
enum API {
    private static let net = NetworkClient.shared

    // MARK: - Auth
    static func login(email: String, password: String) async -> APIResult<TokenResponse> {
        await net.request(path: "/auth/login", method: "POST", body: LoginRequest(email: email, password: password), authenticated: false)
    }

    static func register(email: String, password: String, role: String) async -> APIResult<TokenResponse> {
        await net.request(path: "/auth/register", method: "POST", body: RegisterRequest(email: email, password: password, role: role), authenticated: false)
    }

    static func me() async -> APIResult<UserMe> {
        await net.request(path: "/auth/me")
    }

    // MARK: - Home / Marketplace
    static func home() async -> APIResult<HomeData> {
        await net.request(path: "/stepserve/home", authenticated: false)
    }

    static func categories() async -> APIResult<[Category]> {
        await net.request(path: "/categories", authenticated: false)
    }

    static func services() async -> APIResult<[Service]> {
        await net.request(path: "/services", authenticated: false)
    }

    static func searchServices(query: String = "", categoryId: Int? = nil, location: String = "", minPrice: String = "", maxPrice: String = "") async -> APIResult<[Service]> {
        var params: [String] = []
        if !query.isEmpty    { params.append("query=\(query.urlEncoded)") }
        if let c = categoryId { params.append("category_id=\(c)") }
        if !location.isEmpty { params.append("location=\(location.urlEncoded)") }
        if !minPrice.isEmpty { params.append("min_price=\(minPrice)") }
        if !maxPrice.isEmpty { params.append("max_price=\(maxPrice)") }
        let qs = params.isEmpty ? "" : "?\(params.joined(separator: "&"))"
        return await net.request(path: "/search/services\(qs)", authenticated: false)
    }

    // MARK: - Provider
    static func upsertProfile(_ body: ProviderProfileUpsert) async -> APIResult<ProviderProfile> {
        await net.request(path: "/providers/profile", method: "POST", body: body)
    }

    static func providerDashboard() async -> APIResult<ProviderDashboard> {
        await net.request(path: "/provider/dashboard")
    }

    static func providerListings() async -> APIResult<[Service]> {
        await net.request(path: "/provider/listings")
    }

    static func createService(_ body: ServiceCreate) async -> APIResult<Service> {
        await net.request(path: "/services", method: "POST", body: body)
    }

    static func updateService(id: Int, body: ServiceUpdate) async -> APIResult<Service> {
        await net.request(path: "/services/\(id)", method: "PATCH", body: body)
    }

    static func deleteService(id: Int) async -> APIResult<Service> {
        await net.request(path: "/services/\(id)", method: "DELETE")
    }

    static func payListing(serviceId: Int) async -> APIResult<ListingPaymentResponse> {
        await net.request(path: "/listings/\(serviceId)/pay", method: "POST")
    }

    // MARK: - Documents
    static func providerUploads() async -> APIResult<[ProviderUpload]> {
        await net.request(path: "/providers/uploads")
    }

    static func uploadDocument(data: Data, fileName: String, mimeType: String) async -> APIResult<ProviderUpload> {
        await net.upload(path: "/providers/uploads", fileData: data, fileName: fileName, mimeType: mimeType)
    }

    // MARK: - Bookings
    static func createBooking(_ body: BookingCreate) async -> APIResult<BookingResponse> {
        await net.request(path: "/bookings", method: "POST", body: body)
    }

    static func customerBookings() async -> APIResult<[CustomerBooking]> {
        await net.request(path: "/customer/bookings")
    }

    // MARK: - Payments
    static func createPayment(_ body: PaymentCreate) async -> APIResult<PaymentResponse> {
        await net.request(path: "/payments", method: "POST", body: body)
    }

    // MARK: - Reviews
    static func createReview(_ body: ReviewCreate) async -> APIResult<EmptyResponse> {
        await net.request(path: "/reviews", method: "POST", body: body)
    }

    // MARK: - Admin
    static func adminOverview() async -> APIResult<AdminOverview> {
        await net.request(path: "/admin/overview")
    }

    static func adminUsers() async -> APIResult<[AdminUser]> {
        await net.request(path: "/admin/users")
    }

    static func adminSetUserStatus(userId: Int, active: Bool) async -> APIResult<EmptyResponse> {
        await net.request(path: "/admin/users/\(userId)/status?active=\(active)", method: "PATCH")
    }

    static func adminBookings() async -> APIResult<[AdminBooking]> {
        await net.request(path: "/admin/bookings")
    }

    static func createCategory(_ body: CategoryCreate) async -> APIResult<Category> {
        await net.request(path: "/categories", method: "POST", body: body)
    }
}

// Placeholder for endpoints that return a JSON object we don't need to parse
struct EmptyResponse: Decodable {}

private extension String {
    var urlEncoded: String {
        addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? self
    }
}
